### PR TITLE
Added http background traffic

### PR DESCRIPTION
**Pull request by mskpyn**
Please review my codes, changes have been marked with !! on the comments

### CODE REVIEW

File: sniffles/traffic_writer.py [hunk 1]
```diff
@@ -12,7 +12,7 @@ class TrafficWriter:
         write, call write_packet(length, pkt) where length
         is the length of the packet and pkt is the binary
         string for the packet.  When finished, call close_save_file()
-        to close the file handle.
+        to close the file handl.
```
Comment (msk) [inline sniffles/traffic_writer.py]: Revert this change.

File: sniffles/test/data_files/rules/test_rules1.rules [hunk 2]
```diff
@@ -8,4 +8,4 @@ alert tcp $EXTERNAL_NET any -> $HOME_NET 8081 (msg:"test1-6"; flow:to_server,est
...
 alert tcp $EXTERNAL_NET any -> $HOME_NET $HTTP_PORTS (msg:"test1-8"; flow:to_server,established; content:"Proxy-Connection: Keep-Alive"; http_header; content:"Transfer-Encoding: chunked|0D 0A|Content-Length: 40334"; fast_pattern:only; http_header; content:".dll"; http_uri; classtype:attempted-admin; sid:8;)
 alert tcp $EXTERNAL_NET any -> $HOME_NET 8080 (msg:"test1-9"; flow:to_server,established; content:"%2E%2E/"; fast_pattern:only; content:"%2E%2E/"; nocase; http_raw_uri; pcre:"/\/(\\|%5C)%2E%2E\//"; sid:9;)
-alert tcp $EXTERNAL_NET any -> $HOME_NET $HTTP_PORTS (msg:"test1-10"; flow:to_server,established; content:"POST"; http_method; content:"=|27|"; http_client_body; content:"new"; nocase; http_client_body; pcre:"/new(%20)?(javax?|org|com)/i"; classtype:attempted-admin; sid:10;)
\ No newline at end of file
+alert tcp $EXTERNAL_NET any -> $HOME_NET $HTTP_PORTS (msg:"test1-10"; flow:to_server,established; content:"POST"; http_method; content:"=|27|"; http_client_body; content:"new"; nocase; http_client_body; pcre:"/new(%20)?(javax?|org|com)/i"; classtype:attempted-admin; sid:10;)
```
Comment (vvalgenti) [inline sniffles/test/data_files/rules/test_rules1.rules]: Should have newline at end.

Comment (vvalgenti) [inline sniffles/test/data_files/rules/test_rules1.rules]: I'm not really clear concerning the goal of these changes.  If it is just to generate http traffic, we could setup a petabi rule file or even just use snort rules to generate http traffic as background traffic.  If we want to add something specific, I'd prefer that we not change the parent traffic stream and instead create a child traffic stream object and use that similar to how scan traffic is generated.  These changes are touching an awful lot of code for a relative innocent addition.  I would much prefer if we could make these changes more independent and modular.

File: sniffles/ruletrafficgenerator.py [hunk 3]
```diff
@@ -1600,95 +1642,107 @@ def generate_nfa_data(self, rule=None):
     def generate_http_content(self, rules):
         http_directive_map = {}
         # HTTP/1.1
-        http_text = [72, 84, 84, 80, 47, 49, 46, 49]
+        http_text = [72, 84, 84, 80, 47, 49, 46, 49, 13, 10]
```
Comment (vvalgenti) [inline sniffles/ruletrafficgenerator.py]: Adding cr lf here might break elsewhere.  The current format is supposed to add that separate.  However, that may be an error below.

File: sniffles/snifflesconfig.py [hunk 4]
```diff
@@ -386,12 +395,13 @@ def getWriteRegEx(self):
 
     def setWriteRegExe(self, value):
         self.write_reg_ex = value
-
+
```
Comment (msk) [inline sniffles/snifflesconfig.py]: Do not add space characters to an empty line. They only increase diff without adding any value to the code.

File: sniffles/test/data_files/rules/test_rules1.rules (hunk 2 above)
Comment (msk) [inline sniffles/test/data_files/rules/test_rules1.rules]: Added. Note the icon at the end of the deleted (red) line.

Comment (msk) [inline sniffles/test/data_files/rules/test_rules1.rules]: Thank you. This is the kind of feedbacks I wanted. The intent was not to add a fancy new feature to sniffles, but to let @mskpyn understand the code structure and figure out how to extend sniffles for more functionalities.

As for the goal of these changes, what would you suggest to create a mixture of a large number of HTTP streams, of which the payloads may be described by numerous regular expressions or other models? What would be the best way to create a group of random streams without defining each stream individually?

Comment (vvalgenti) [inline sniffles/test/data_files/rules/test_rules1.rules]: The intended way would be to create an HTTP_Stream that extends Traffic_Stream and wrap all necessary functionality in that.  You can use the ScanTraffic traffic stream and its use as a guide.  For the http generation itself, I think you can just use what is there without the need to change any of it (i.e. you can feed it a rule or set of rules that meets your criteria without every having to actually have rules--in other words you could create the rules in the HTTP_Stream object).  I am not saying that it would be easier, but it would be more modular.

The second choice, which should have a user story somewhere, would be to create a script to build rules.  Use that to create a rule-set that meets criteria and then Sniffles doesn't need to change at all.  I know I want a tool like that eventually so that we can have more control over what is generated.  It helps in many evaluation cases.  However, it may be a bit much for this particular case.  Also, to note, a Rule can does not need to define a specific stream.  A single rule can define an arbitrary stream and use a regex as a content generator.  It can define number of packets, etc, but best of all, can be used over and over to create multiple, different requests.

Comment (mskpyn) [inline sniffles/test/data_files/rules/test_rules1.rules]: Thank you for the thorough feedback! I've reverted the source codes to its original states and been working on sniffles to create background traffic in a modular and independent way.

File: sniffles/ruletrafficgenerator.py [hunk 5]
```diff
@@ -985,6 +985,58 @@ def updateSequence(self, dir="to server", data_len=1):
...
+        self.sport = Port(rule.getSport())
+        self.dport = rule.getDport()
+        self.rule = rule
+        self.rand = False
+        self.current_seq_a_to_b = random.randint(0, 4000000000)
+        self.current_ack_a_to_b = 0
+        self.current_seq_b_to_a = random.randint(0, 4000000000)
+        self.current_ack_b_to_a = 0
+
+        if sconf:
```
Comment (vvalgenti) [inline sniffles/ruletrafficgenerator.py]: If you pass in the sconf, then I don't think that need to duplicate these values.

File: sniffles/ruletrafficgenerator.py [hunk 6]
```diff
@@ -985,6 +985,58 @@ def updateSequence(self, dir="to server", data_len=1):
...
+            if sconf.getLatency() > 0:
+                self.latency = sconf.getLatency()
+            self.mac_def_file = sconf.getMacAddrDef()
+            if self.mac_def_file:
+                self.mac_gen = ETHERNET_HDR_GEN_DISTRIBUTION
+            self.pkt_len = sconf.getPktLength()
+            if sconf.getPktsPerStream() > 1:
+                self.packets_in_stream = sconf.getPktsPerStream()
+
+        if ipv6_percent > 0:
```
Comment (vvalgenti) [inline sniffles/ruletrafficgenerator.py]: As above, if you have passed in the sconf, I don't believe you need to worry about this either.  It will be handled in the super constuctor.

File: sniffles/snifflesconfig.py [hunk 7]
```diff
@@ -648,6 +665,8 @@ def usage(self):
         print("   sent.  Off by default.")
         print("-b Bi-directional data: Send data in both directions.")
         print("   Off by default.  Automatically sets TCP acks.")
+        # !! Usage for background Traffic
+        print("-B Background Traffic: Percentage of Background traffic to create")
```
Comment (vvalgenti) [inline sniffles/snifflesconfig.py]: End with a period.  Also, provide some clue as to what the background traffic will contain (i.e. generic http and ftp traffic).

**Comment by vvalgenti** [comment]
Please update the README documenting the background traffic.  Also, please provide an example in the README.  Finally, please add test cases.  Otherwise, looks much better.
